Add fee collectors management how-to (#1150)
OffchainLabs/arbitrum-docs: arbitrum-docs/launch-orbit-chain/how-tos/manage-fee-collectors.md

---
title: 'How to manage the fee collector addresses of your Orbit chain'
description: 'Learn how to manage the fee collector addresses of your Orbit chain'
author: jose-franco
sme: jose-franco
sidebar_position: 6
content_type: how-to
---

import PublicPreviewBannerPartial from '../partials/_orbit-public-preview-banner-partial.md';

<PublicPreviewBannerPartial />

> As part of the activity of an Orbit chain, different fees are collected with every transaction. All these fees are collected as a single amount (the transaction fees), but are internally split into different components depending on their purpose. Each component can also be transferred to a different fee collector address that can be configured on your chain.

Reviewer: ```suggestion
As part of the activity of an Orbit chain, different fees are collected with every transaction. These fees are collected as a single amount (the transaction fees) but are internally split into different components depending on their purpose. Each component can also be transferred to a different fee collector address that can be configured on your chain.
```

> This guide describes the different fees that are collected, and explains how to specify the fee collector address on your chain for each specific fee type.

Reviewer: ```suggestion
This guide describes the different fees that are collected, and explains how to specify the fee collector address on your chain for each fee type.
```

## What fees are collected on an Orbit chain?

> There are 4 fee types that are collected on every transaction of an Orbit chain:

Reviewer: ```suggestion
There are four fee types that are collected on every transaction of an Orbit chain:
```

> - **Orbit base fee**: fees paid for the execution of the transaction on the chain based on the minimum base price configured.

Reviewer: ```suggestion
- **Orbit base fee**: fees paid for executing transactions on the chain based on the minimum base price configured.
```

> - **Orbit surplus fee**: if the chain is congested (i.e., the base price paid for the transaction is higher than the minimum base price), these fees account for the execution of the transaction on the chain based on any gas price paid above the minimum base price configured.

Reviewer: ```suggestion
- **Orbit surplus fee**: if the chain is congested (i.e., the base price paid for the transaction is higher than the minimum base price), these fees account for executing transactions on the chain based on any gas price paid above the minimum base price configured.
```

> - **Parent chain base fee**: relative fees paid for posting the transaction on the parent chain. This is calculated based on the transaction's estimated size and the current view of the parent chain's base fee.

Reviewer: ```suggestion
- **Parent chain base fee**: relative fees paid for posting the transaction on the parent chain. The amount is calculated based on the transaction's estimated size and the current view of the parent chains' base fee.
```

- **Parent chain surplus fee**: if configured, these are extra fees rewarded to the batch poster.

You can find more detailed information about these fee types in these pages:

- [L2 fees](/arbos/gas.mdx) for the Orbit base fee and surplus fee
- [L1 fees](/arbos/l1-pricing.mdx) for the Parent chain base fee and surplus fee

## How to configure the fee collector addresses?

> Let's now look at how to configure the collector addresses for each of the fee types.

Reviewer: ```suggestion
Let's now look at how to configure the collector addresses for each fee type.
```

### Orbit base fee

> Orbit base fees are paid to the `infraFeeAccount` configured in your chain. You can get the current address configured, by calling the method `getInfraFeeAccount()` of the [ArbOwnerPublic](/build-decentralized-apps/precompiles/02-reference.md#arbownerpublic) precompile. For example:

Reviewer: ```suggestion
Orbit base fees are paid to the infraFeeAccount configured in your chain. You can retrieve the current address by calling the getInfraFeeAccount() method of the [ArbOwnerPublic](https://arbitrum-docs-igdj6umeq-offchain-labs.vercel.app/build-decentralized-apps/precompiles/reference#arbownerpublic) precompile. For example:
```

    chain: <OrbitChainDefinition>,
    transport: http(),
}).extend(arbOwnerPublicActions);

const infraFeeAccount = await orbitChainClient.arbOwnerReadContract({
    functionName: 'getInfraFeeAccount',
});
```

To set a new `infraFeeAccount`, use the method `setInfraFeeAccount(address)` of the [ArbOwner](/build-decentralized-apps/precompiles/02-reference.md#arbowner) precompile. For example:

```shell
cast send --rpc-url $ORBIT_CHAIN_RPC --private-key $OWNER_PRIVATE_KEY 0x0000000000000000000000000000000000000070 "setInfraFeeAccount(address) ()" $NEW_INFRAFEEACCOUNT_ADDRESS
```

Or using the Orbit SDK:

```typescript
const owner = privateKeyToAccount(<OwnerPrivateKey>);
const orbitChainClient = createPublicClient({
    chain: <OrbitChainDefinition>,
    transport: http(),
}).extend(arbOwnerPublicActions);

const transactionRequest = await orbitChainClient.arbOwnerPrepareTransactionRequest({
    functionName: 'setInfraFeeAccount',
    args: [<NewInfraFeeAccountAddress>],
    upgradeExecutor: false,
    account: owner.address,
});

await orbitChainClient.sendRawTransaction({
    serializedTransaction: await owner.signTransaction(transactionRequest),
});
```

### Orbit surplus fee

> Orbit surplus fees are paid to the `networkFeeAccount` configured in your chain. You can get the current address configured, by calling the method `getNetworkFeeAccount()` of the [ArbOwnerPublic](/build-decentralized-apps/precompiles/02-reference.md#arbownerpublic) precompile. For example:

Reviewer: ```suggestion
Orbit surplus fees are paid to the `networkFeeAccount` configured in your chain. You can retrieve the current address by calling the method `getNetworkFeeAccount()` of the [ArbOwnerPublic](/build-decentralized-apps/precompiles/02-reference.md#arbownerpublic) precompile. For example:
```

```shell
cast call --rpc-url $ORBIT_CHAIN_RPC 0x000000000000000000000000000000000000006B "getNetworkFeeAccount() (address)"
```

_Note: The [ArbOwner](/build-decentralized-apps/precompiles/02-reference.md#arbowner) precompile also has a `getNetworkFeeAccount()` method that can be used, but only by the owner of the chain._

> Alternatively, you can use the Orbit SDK to get the current address configured as networkFeeAccount, by calling the [ArbOwner](/build-decentralized-apps/precompiles/02-reference.md#arbowner) precompile:

Reviewer: ```suggestion
Alternatively, you can use the Orbit SDK to retrieve the current address as `networkFeeAccount`, by calling the [ArbOwner](/build-decentralized-apps/precompiles/02-reference.md#arbowner) precompile:
```

const networkFeeAccount = await orbitChainClient.arbOwnerReadContract({
    functionName: 'getNetworkFeeAccount',
});
```

To set a new `networkFeeAccount`, use the method `setNetworkFeeAccount(address)` of the [ArbOwner](/build-decentralized-apps/precompiles/02-reference.md#arbowner) precompile. For example:

```shell
cast send --rpc-url $ORBIT_CHAIN_RPC --private-key $OWNER_PRIVATE_KEY 0x0000000000000000000000000000000000000070 "setNetworkFeeAccount(address) ()" $NEW_NETWORKFEEACCOUNT_ADDRESS
```

Or using the Orbit SDK:

```typescript
const owner = privateKeyToAccount(<OwnerPrivateKey>);
const orbitChainClient = createPublicClient({
    chain: <OrbitChainDefinition>,
    transport: http(),
}).extend(arbOwnerPublicActions);

const transactionRequest = await orbitChainClient.arbOwnerPrepareTransactionRequest({
    functionName: 'setNetworkFeeAccount',
    args: [<NewNetworkFeeAccountAddress>],
    upgradeExecutor: false,
    account: owner.address,
});

await orbitChainClient.sendRawTransaction({
    serializedTransaction: await owner.signTransaction(transactionRequest),
});
```

### Parent chain base fee

:::info ArbAggregator currently not supported in the Orbit SDK

> Reading information from the ArgAggregator precompile, or using it to set new information is currently not supported in the Orbit SDK, but will soon be added. So for now, this subsection will only show examples using `cast call` and `cast send`.

Reviewer: ```suggestion
Reading information from the `ArgAggregator` precompile or using it to set new information is currently not supported in the Orbit SDK but will be added soon. So, for now, this subsection will only show examples using `cast call` and `cast send`.
```

:::

Parent chain base fees are paid to the fee collector of the active batch poster configured in your chain. The current configured batch posters can be obtained by calling the method `getBatchPosters()` of the [ArbAggregator](/build-decentralized-apps/precompiles/02-reference.md#arbaggregator) precompile. For example:

```shell
cast call --rpc-url $ORBIT_CHAIN_RPC 0x000000000000000000000000000000000000006D "getBatchPosters() (address[])"
```

This list has to also be verified against the `SequencerInbox` contract living on the parent chain. This contract needs to have any of those addresses in its `isBatchPoster` mapping. To verify a specific address, you can check the mapping directly like this:

```shell
cast call --rpc-url $PARENT_CHAIN_RPC $SEQUENCER_INBOX_ADDRESS "isBatchPoster(address) (bool)" $BATCH_POSTER_ADDRESS
```

> Once we have the current batch poster, we can obtain the fee collector address configured for that batch poster by calling the method `getFeeCollector(address)` of the [ArbAggregator](/build-decentralized-apps/precompiles/02-reference.md#arbaggregator) precompile and passing the address of the batch poster.

Reviewer: ```suggestion
Once you have the current batch poster, you can obtain the fee collector address configured for that batch poster by calling the method `getFeeCollector(address)` of the [ArbAggregator](/build-decentralized-apps/precompiles/02-reference.md#arbaggregator) precompile and passing the address of the batch poster.
```

});

await orbitChainClient.sendRawTransaction({
    serializedTransaction: await owner.signTransaction(transactionRequest),
});
```

To change the reward rate, you can use the method `setL1PricingRewardRate(uint64)` of the [ArbOwner](/build-decentralized-apps/precompiles/02-reference.md#arbowner) precompile and pass the amount of wei per gas unit to reward. For example:

```shell
cast send --rpc-url $ORBIT_CHAIN_RPC --private-key $OWNER_PRIVATE_KEY 0x0000000000000000000000000000000000000070 "setL1PricingRewardRate(uint64) ()" $NEW_REWARD_RATE
```

Or using the Orbit SDK:

```typescript
const owner = privateKeyToAccount(<OwnerPrivateKey>);
const orbitChainClient = createPublicClient({
    chain: <OrbitChainDefinition>,
    transport: http(),
}).extend(arbOwnerPublicActions);

const transactionRequest = await orbitChainClient.arbOwnerPrepareTransactionRequest({
    functionName: 'setL1PricingRewardRate',
    args: [<NewRewardRate>],
    upgradeExecutor: false,
    account: owner.address,
});

await orbitChainClient.sendRawTransaction({
    serializedTransaction: await owner.signTransaction(transactionRequest),
});
```

## How to use the fee distribution contracts?

> For now, we've described how to set the individual collector addresses for each of the fee types. Some chains may require to have multiple addresses receive the collected fees of any of the available types. For those cases, there's the possibility of using a distributor contract that can gather all fees of a specific type and distribute those among multiple addresses.

Reviewer: ```suggestion
For now, we've described how to set the individual collector addresses for each fee type. Some chains may require multiple addresses to receive the collected fees of any of the available types. In those cases, there's the possibility of using a distributor contract that can gather all fees of a specific type and distribute those among multiple addresses.
```

This section shows how to configure a distributor contract to manage the fees of a specific type.

:::info Distributor contracts currently not supported in the Orbit SDK

> Currently, the Orbit SDK doesn't support deploying and configuring distribution contracts, but will soon be added. So for now, this section will only show examples using `cast send`.

Reviewer: ```suggestion
Currently, the Orbit SDK doesn't support deploying and configuring distribution contracts, but it will soon be added. So, for now, this section will only show examples using `cast send`.
```

:::

### Step 1. Deploy the distributor contract

An example implementation of a distributor contract can be found [here](https://github.com/OffchainLabs/fund-distribution-contracts/blob/main/src/RewardDistributor.sol). You'll have to deploy this contract on your Orbit chain.

### Step 2. Set the contract address as the desired fee type collector address

Use the instructions provided in the previous section to set the address of the deployed distributor contract as the collector of the desired fee type. For example, if you want the distributor contract to manage the Orbit surplus fees, set the `networkFeeAccount` to the address of the deployed contract.

### Step 3. Configure the recipients of fees in the contract

Now you can set the different addresses that will be receiving fees from that distributor contract. To do that, you can call the method `setRecipients(address[], uint256[])` of the distributor contract, and specify the list of addresses that will be receiving fees, and the proportion of fees for each address.

> For example, if you want to set 2 addresses as receivers, with the first one receiving 80% of the fees, and the second one receiving 20% of the fees, you'll use the following parameters:

Reviewer: ```suggestion
For example, if you want to set two addresses as receivers, with the first one receiving 80% of the fees and the second one receiving 20% of the fees, you'll use the following parameters:
```
